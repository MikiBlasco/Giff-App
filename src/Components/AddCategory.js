import React, {useState} from 'react';
import PropTypes from 'prop-types';


const AddCategory = (props) => {

   const [inputValue, setInputValue] = useState(''); //ponemos " " por que si dejamos vacío es undefined. Luego el del inputvalue en value es también undefined. Entonces nos sale el Warning the "component changing an uncontrolled input..."

   const handleInputChange = (e) => {
       setInputValue(e.target.value)
   }

   const handleSubmit = (e) => {
       e.preventDefault(); 

       //validation (trim para borrar espacios en blanco de de los lados)
       if ( inputValue.trim().length > 2 ) {

           props.setCategories(cat => [inputValue, ...cat]) 
           setInputValue('') // para que se vacíe el placeHolder al hacer submit. 

         }           
   }
  
    return (

        <form onSubmit={handleSubmit}>
            <input 
                type="text"
                value={inputValue}
                onChange={ handleInputChange }
            />
        </form>

    )

}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}

export default AddCategory