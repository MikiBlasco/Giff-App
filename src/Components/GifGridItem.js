import React from 'react'
import ProptTypes from 'prop-types'

export const GifGridItem = ({title, url}) => {

  return (

    <div className='card animate__animated animate__bounceIn'>

    <img src={ url } alt={ title } />
    <p> { title } </p>  

    </div>
  )

}

GifGridItem.proptTypes = {
  title: ProptTypes.string.isRequired,
  url: ProptTypes.string.isRequired
}




 
