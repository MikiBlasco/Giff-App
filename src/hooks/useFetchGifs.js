import { useState, useEffect } from 'react' //hace falta porque estos hooks tb pueden tener su estado y podría indicarle a otros componenetes que lo usen que RENDERICEN porque algo cambió 
import { getFetchGifs } from '../helpers/getFetchGifs';

export const useFetchGifs = ( category ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {

        getFetchGifs( category )
          .then( imgs => { 

                 setState({
                     data: imgs,
                     loading: false,
               }); 

        });
        
    }, [category] )
        
            // useEffect( () => {
    //     //le mandamos "category" a la func que está en GifGrid.
    //     //como GifGrid retorna una promesa le podemos poner el"then".
    //     getFetchGifs(category)
    //     //el then contendra en el capuccino "imgs" el return/resol. de la promesa (con la colección de gifs) que se lo meteremos a la función de setImages.
    //     .then( imgs => setImages( imgs ) )
    //     //.then( setImages ) así sería lo mismo porque coincide el capuccino de then con lo que meteremos en setImages.

    // }, [ category ] )

    


    return state; // { data:[], loading: true };
}
