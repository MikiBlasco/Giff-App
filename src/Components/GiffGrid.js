import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GiffGrid = ({ category }) => {
    //renombramos con la desestructuración la data para que se llame imagenes.
    const {data: images, loading} = useFetchGifs( category ); //state es producto del customHook


  return (
    <>

            <h3 className='animate__animated animate__fadeIn'>{category}</h3>

            { loading && <p className='animate__animated animate__flash'>Loading...</p> }

            <div className='card-grid'>


                {
                      images.map((img) => 
                        <GifGridItem 
                          key={ img.id }    
                          { ...img } //así mando cada una de las propiedades del objeto "img" como si fueran indendientes.
                          /> )
                }
            
            </div>

    </>

  )
}
