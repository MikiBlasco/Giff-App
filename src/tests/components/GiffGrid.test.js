import React from 'react';
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import { GiffGrid } from '../../Components/GiffGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';

jest.mock('../../hooks/useFetchGifs') //As√≠ simulo cualquier llamada a este archivo y suponer/controlar la info que esto va a responder.


describe('pruebas del componente <GiffGrid/>', () => { 
    
    test('should match with snapshot', () => { 

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        })


        const wrapper = shallow(<GiffGrid category='category' />);
        expect(wrapper).toMatchSnapshot();
     })

    test('should show items when images are loaded with the customhook to fetch', () => { 

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'cualquier cosa'
        }]

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        })

        const wrapper = shallow(<GiffGrid category='category' />);

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GifGridItem').length ).toBe(gifs.length);
        
     })


 })