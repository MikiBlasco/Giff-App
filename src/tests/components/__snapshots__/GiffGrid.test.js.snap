// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas del componente <GiffGrid/> should match with snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    category
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`pruebas del componente <GiffGrid/> should show items when images are loaded with the customhook to fetch 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    category
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;
